File: day03/part2.nim
import os
import sequtils
import strutils


proc compute(data: seq[string]): int =
  const slopes = [(1, 1), (1, 3), (1, 5), (1, 7), (2, 1)]

  let
    width = len data[0]
    height = len data
  var
    product = 1

  for dy, dx in slopes.items:
    proc check(y: int): bool =
      data[y][(y div dy) * dx mod width] == '#'

    product *= max(
      1,
      toSeq(countUp(0, height - 1, dy))
      .filter(check)
      .len()
    )

  product


proc main() =
  let data = readFile(os.joinPath(os.getAppDir(), "input.txt"))
  echo compute(data.strip().splitLines(false))


if isMainModule:
  main()

File: day03/part2.py
import itertools

_TESTCASE = """\
..##.......
#...#...#..
.#....#..#.
..#.#...#.#
.#...##..#.
..#.##.....
.#.#.#....#
.#........#
#.##...#...
#...##....#
.#..#...#.#
""".splitlines()


def compute(data):
    """
    >>> compute(_TESTCASE)
    336
    """
    lines = list(data)
    width, height = len(lines[0]), len(lines)
    slopes = [(1, 1), (1, 3), (1, 5), (1, 7), (2, 1)]
    product = 1

    for dy, dx in slopes:
        coords = itertools.takewhile(
            lambda coord: coord[0] < height,
            map(lambda n: (n * dy, n * dx % width), itertools.count()),
        )
        product *= max(1, sum(lines[y][x] == "#" for y, x in coords))

    return product


def main():
    import pathlib

    input_path = pathlib.Path(__file__).with_name("input.txt")
    # `splitlines` removes trailing "\n" whereas iter(file.open()) would not
    print(compute(input_path.read_text().splitlines()))


if __name__ == "__main__":
    main()
